### DISCUSSION TITLE
Prisma, Transactions and Software architecture

### DISCUSSION THREAD

**Discussion by mistyharsh** [prisma/prisma#7754] [Q&A]
Hi Team,

Thanks for building the wonderful tool. The experience has been amazing so far. This is not really an issue but a general commentary and concern after using Prisma for a while.

### What I am building?
[Bisa](https://github.com/RanoIO/bisa) - It is an experiment in content platform for community owned infrastructure. The API exposed is GraphQL using Nexus.js and database is PostgreSQL powered by Prisma for connectivity.

### Problem description
While I certainly understand how Prisma intends us to consume transactions as described [here](https://www.prisma.io/docs/concepts/components/prisma-client/transactions), [here](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide) and [here](https://www.prisma.io/blog/how-prisma-supports-transactions-x45s1d5l0ww1). In general, it works well.

This current transaction API is not sufficient to build robust concurrent applications. It is not because of the mindset but because of commitment to follow certain software architectural attributes that have worked for decades. In general, when building a back-end, we divide it into layers - either onions or horizontal layers. In my case, I try to rely on some form of [Hexagonal architecture](https://fsharpforfunandprofit.com/posts/recipe-part3/) even if not 100% perfect with TypeScript/Node.js. Either way, we have a business logic written in BAL (Business access layer) or Domain layer. While persistence is not a Domain layer concern and Prisma is doing excellent job with it. It nicely abstracts away the complexities and provide proper leak-proof persistence layer.

However, transaction (classical read-modify-commit) are almost always a domain layer concern. It is a part of business logic and as such this layer should be in charge of controlling transaction lifecycle. And, that's where the limitations with current APIs become very problematic. So, what's happening here is Prisma is trying to own business logic concerns in order to simplify the data access.

### Practical example
Consider following model in Bisa.

```prisma
model Quota {
  id              BigInt    @id
  project         Project   @relation(fields: [id], references: [id])
  sizeInMB        Int
  staffCapacity   Int

  // Added extra field to use OCC - Optimistic concurrency control
  occupied        Int
}

model Invitation {
  id            BigInt      @id @default(autoincrement())
  code          String      @unique

  firstName     String
  lastName      String
  email         String

  project       Project     @relation(fields: [projectId], references: [id])
  projectId     BigInt

  createdAt     DateTime    @db.Timestamptz() @default(now())
  updatedAt     DateTime    @db.Timestamptz() @updatedAt

  @@unique([email, projectId])
}
```
Here I want to send invitations not exceeding the `staffCapacity` as mentioned by the the `Quota` entity. In traditional world, I would do that in **Read-Edit-Commit** cycle as follows:
 - [Domain Layer] - Start a transaction (serializable).
 - [DAL] - Write a function to get said `Entity` using repository pattern or some function from DAL layer.
 - [DAL] - Write another function to get count of issued invitations.
 - [Domain] Validate and generate the invitation if the count doesn't exceed.
 - [Domain] Commit a transaction.
 - [Domain] Handle transaction error.

Note here that architectural boundaries are very well-defined and preserved. Now with Prisma, I can try to employ Optimistic Concurrency Control to manage this transaction. So, as a version token, I add `occupied` field to `Quota` (It would serve like a version equivalent) which I can use during nested update like this:

```typescript

// DAL
const quota = await db.quota.findFirst({ where: { id: BigInt(1) } });

// Domain - business logic
if (!quota) {
  return R.ofError(ErrorCode.NOT_FOUND, 'Quota information not found');
}

// Domain - business logic
if (quota.occupied >= quota.staffCapacity) {
  return R.ofError(ErrorCode.QUOTA_FULL, 'Staff capacity if full. New staff cannot be added');
}

// DAL
db.quota.update({
  data: {
    project: {
      update: {
        invitations: {
          create: {
            code: "",
            email: "",
            firstName: "",
            lastName: ""
          }
        }
      }
    }
  },
  where: {
    id: 1,

    // PROBLEM! NOT SUPPORTED!
    occupied: quota.occupied
  }
});
```

The above example should solve the problem but `where` clause doesn't support filtering on the `occupied` field as it is not indexed. I can create an unique index for `@@unique([id, occupied])` and then `where` clause can accept that tuple to filter. But that is not clean solution. Why a data modelling engineer who might be a separate person to design schema would set index for `occupied` field. That is leaking the abstraction at a database level. It is not even part of integrity to do so at DB level. Alternately, Prisma's `updateMany` clause support complex filters but doesn't provide nested updates which is a blocker to address it.

So, finally, I can execute the raw queries using query builder and manually mange it but that really defeats the purpose of using Prisma. Additionally, the DAL and Domain code is interleaved in a single function.

### Expectation
I would like to preserve the architectural boundaries while building the application with Prisma. Persistence and Transaction are two different things. If this is not how it is intended it to be, what is the alternate way to architecture for large-scale applications with Prisma as a Data Access Layer? How to segregate between business logic and data access?

Apologies for long thread but I am thinking hard and yet I seem to not find a clean solution that respects the software architectural boundaries.

**Comment by ryands17** [comment]
@mistyharsh 👋 

In this case, the best way to do this would be via Optimistic Concurrency Control with a version number as Prisma doesn't support long running transactions as of now. There's a request for that same here that we have started work on and will take approximately three months to ship:

https://github.com/prisma/prisma/issues/1844#issuecomment-846477520

As for the Hexagonal pattern, this isn't directly possible as Prisma isn't a traditional ORM where you would create classes and setters along with the required helper methods. In this case, you would need to opt for a simple **Controller -> Routes -> Result** pattern where you write your logic on the controllers directly that include the relevant Prisma logic (which can be encapsulated within functions) and then use these in your routes where your are returning them.

Let me know if you have any further questions :)

**Reply by mistyharsh** [reply]
Thank you @ryands17 for the help.

I ended up doing the OCC. And, yes, I understand the it is not hexagonal pattern and probably meaningless. But in general, even without Hexagonal architecture, I would like to keep my query functions separate from business logic as it gets in a way of reading the code. It further becomes more problematic with scenarios where I need to re-use the queries across multiple domain layers functions.

For [Bisa][1], I believe I seem to have achieved a clean segregation between queries and business logic. Main problem was:
  - I wanted to create a user (query 1) and delete its invitation (query 2). Both should be defined in a `data` layer where all the queries reside.
  - This must run as a transaction using Prisma `$transaction` API. But the challenge here is that user creation is an `async` function (example below) as the password hashing returns a Promise. So, I lost the `PrismaPromise<T>` when returning from the user creation function.
  - I ended up wrapping the query as a callback function (Call it a **lazy promise** or **lazy query** if you will) whose control would then be given to the caller as shown below. The caller would make sure that it is called in the **same time tick** to allow the use of `$transaction` API.

```typescript

// This is a function in the `data` layer.
export async function createAdminWithNewUser(db: PrismaClient, input: UserInput, password: string) {

  const userPayload = await buildNewUser(input, password);

  // Notice the use of Callback
  return () =>
    db.admin.create({
      data: {
        superAdmin: false,
        user: { create: userPayload }
      }
    });
}

// Another function from `data` layer that returns `PrismaPromise` instance
export function deleteInvitation(db: PrismaClient, invitationId: bigint) {
  return db.invitation.delete({
    where: { id: invitationId }
  });
}


// Usage - This is a function from the `domain` layer that uses above two 
async function createAdmin(ctx: Context, invitation: Invitation, password: string) {

  const { db } = ctx;

  try {
    // Instead of returning Promise<PrismaQuery>,
    //  return () => PrismaPromise<PrismaQuery>
    const requestFn = await createAdminWithNewUser(db, invitation, password);

    const request = requestFn();
    const cleanup = deleteInvitation(db, invitation.id);

    await db.$transaction([request, cleanup]);

    return R.of({ status: true });

  } catch (error) {
    // TODO: Currently, it is not possible to cleanly map the errors occurred during $transaction.
    return R.of({ status: false });
  }
}
```

I am currently happy with this approach and code looks readable enough. I am using this `createAdminWithNewUser` function from multiple top-level domain layer functions. I don't want to call it that, but my query functions are ending up like a monadic computation (lazier once) where you have all the per-requisites ready for any given query and the called in the end simply wrapping up in a transaction. What I am losing here is automatic type inference but that's a TS issue and nothing to do with Prisma. But more verbose code is acceptable for the readability & separation it is providing.

What do you think about this? Do you foresee any problem with this approach?

[1]: https://github.com/RanoIO/bisa

**Reply by ryands17** [reply]
Looks great!
I don't think you should lose type inference. Which part is that happening exactly?

**Reply by mistyharsh** [reply]
It is not about **loosing** the static type safety. It is more like becoming verbose and loosing **automatic** type inference.

**Comment by sorenbs** [comment]
This is the tracking issue for specific support for OCC: https://github.com/prisma/prisma/issues/4988

**Reply by sorenbs** [reply]
@mistyharsh - would introducing the `if` construct to enable OCC fully solve your issue?

It would look like this:

```
// DAL
const quota = await db.quota.findFirst({ where: { id: BigInt(1) } });

// Domain - business logic
if (!quota) {
  return R.ofError(ErrorCode.NOT_FOUND, 'Quota information not found');
}

// Domain - business logic
if (quota.occupied >= quota.staffCapacity) {
  return R.ofError(ErrorCode.QUOTA_FULL, 'Staff capacity if full. New staff cannot be added');
}

// DAL
db.quota.update({
  data: {
    project: {
      update: {
        invitations: {
          create: {
            code: "",
            email: "",
            firstName: "",
            lastName: ""
          }
        }
      }
    }
  },
  where: {
    id: 1,

  },
  if: {

    // THIS NOW WORKS
    occupied: quota.occupied
  }
});
```

**Reply by matthewmueller** [reply]
In the meantime, if you'd like to implement OCC today, you can use an auto-incrementing version. You can find an example in our [Transaction Guide](https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide#optimistic-concurrency-control).

**Reply by mistyharsh** [reply]
@sorenbs I like the idea of `if` but somehow against the grains of SQL. Isn't `where` clause supposed to do the same thing? Also, I am not sure it would be generic enough.

@matthewmueller I ended up building OCC as the guide you shared but with a slight twist. Here I don't have pre-generated **seat** that can be claimed as it is shown in docs. The invitation in my case is generated on the fly. In fact, in ideal world, there should be no need for `occupied` column as it is an derived value which I simply dislike when my persistence is relational DB. But for now, I can live with it to achieve OCC.

I ended up adding a `check` constraint on my DB for full proofing - 

    -- This is an empty migration.
    ALTER TABLE "quota"
      ADD CONSTRAINT "max_capacity" CHECK ("occupied" <= "staff_capacity");

But in general, I feel like this is a business logic which should stay in my app code. And, only way to do that is using **LRT - Long Running Transaction** (Again not really Long Running but just explicitly developer controlled).

To summarize, I have two problems here:
  - Having the need to add a derived value of `occupied` column which is not necessary. In general, I would use [SQL count as shown here][1].
  - Do not leak core business rule in the database (the `occupied` less than `staff_capacity`)

Traditionally, **transaction** would solve this problem completely in the Business Layer addressing both the concerns.

Having said that, I am happy with the Prisma way of doing thing and let us if this is scalable enough across large schema and developers. Time would tell.

  [1]: https://stackoverflow.com/a/26448803/5723098

**Reply by sorenbs** [reply]
> @sorenbs I like the idea of `if` but somehow against the grains of SQL. Isn't where clause supposed to do the same thing?

This is an interesting question. My view is that SQL is overly generic. For example - the difference between deleting a user who is closing their account and deleting all users in the database is just a few characters. Clearly, the intend is very different in the two cases, and a language designed today taking into account modern software development practices would clearly distinguish the two.

Similarly, there is a difference in intend between "find the record I want to update" and "but don't update it in case of X". And this is the reason we will introduce a separate `if` keyword.

If you are okay with a more generic API, you could probably achieve this already by using [`updateMany`](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#updatemany) as it allows filtering on non-unique attributes.

**Reply by sorenbs** [reply]
@mistyharsh - I am also curious. Would you think it was acceptable to add a constraint like that in the Prisma Schema and have Prisma enforce it in all inserts and updates using [this pattern](https://github.com/prisma/prisma/issues/1844#issuecomment-830860673)?

Maybe the Prisma Schema could look like this:

```prisma
model Quota {
  id              BigInt    @id
  project         Project   @relation(fields: [id], references: [id])
  sizeInMB        Int
  staffCapacity   Int

  // Added extra field to use OCC - Optimistic concurrency control
  occupied        Int

  @@constraint(occupied <= staffCapacity)
}
```

**Reply by mistyharsh** [reply]
@sorenbs I am blown away by both the approaches. They both make sense. As far as the `if` is concerned, more experience folks should pitch in. I am not really sure. But this constraint approach `@@constraint` is indeed awesome and a killer. It might need more extensive work on designing the generic grammer/DSL. There many ways how constraints are defined. It should sufficiently be able to cover most common cases to start with.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
